sdio/imx6ull/imx6ull-sdio.c

```diff
@@ -0,0 +1,871 @@
+{
+	uint32_t val, cmdFrame;
+
+	if (res)
+		*res = 0;
+
+	val = *(sdio_common.base + reg_pres_state);
+	if (val & 0x7) {
+		printf("libsdio: data interface is busy (pres_state=%x)\n", val);
+		return -1;
+	}
+
+	cmdFrame = (cmd & 0x3f) << 24;
+
+	/* set response type */
+	switch (cmd) {
+		case SDIO_CMD_STATE_IDLE:
+			/* no response */
+			break;
+		case SDIO_CMD_RELATIVE_ADDR:
+			cmdFrame |= 1 << 20;   /* CICEN=1 */
+			cmdFrame |= 1 << 19;   /* CCCEN=1 */
+			cmdFrame |= 0x2 << 16; /* RSPTYP=2 */
+			break;
+		case SDIO_CMD_SEND_OP_COND:
+			cmdFrame |= 0x2 << 16; /* RSPTYP=2 */
+			break;
+		case SDIO_CMD_SELECT_CARD:
+			cmdFrame |= 1 << 20;   /* CICEN=1 */
+			cmdFrame |= 1 << 19;   /* CCCEN=1 */
+			cmdFrame |= 0x3 << 16; /* RSPTYP=3 */
+			break;
+		case SDIO_CMD_STATE_INACTIVE:
+			/* no response */
+			break;
+		case SDIO_CMD_RW_DIRECT:
+			cmdFrame |= 1 << 20;   /* CICEN=1 */
+			cmdFrame |= 1 << 19;   /* CCCEN=1 */
+			cmdFrame |= 0x2 << 16; /* RSPTYP=2 */
+			break;
+		default:
+			return -1;
+	}
+
+	*(sdio_common.base + reg_mix_ctrl) = 1U << 31;
+	*(sdio_common.base + reg_blk_att) = 0;
+	*(sdio_common.base + reg_cmd_arg) = arg;
+	*(sdio_common.base + reg_cmd_xfer_typ) = cmdFrame;
+
+	/* wait 1 ms max */
+	if (wait_for_cmd(0x1, 10, 100) < 0) {
+		reset_cmd_block();
+		return -1;
+	}
+
+	/* clear status flags */
+	*(sdio_common.base + reg_int_status) = 0x1; /* CC=1 */
+
+	/* retrieve response */
+	switch (cmd) {
+		case SDIO_CMD_STATE_IDLE:
+			/* no response */
+			break;
+		case SDIO_CMD_RELATIVE_ADDR:
+			val = *(sdio_common.base + reg_cmd_rsp0);
+			break;
+		case SDIO_CMD_SEND_OP_COND:
+			val = *(sdio_common.base + reg_cmd_rsp0);
+			break;
+		case SDIO_CMD_SELECT_CARD:
+			val = *(sdio_common.base + reg_cmd_rsp3);
+			break;
+		case SDIO_CMD_STATE_INACTIVE:
+			/* no response */
+			break;
+		case SDIO_CMD_RW_DIRECT:
+			val = *(sdio_common.base + reg_cmd_rsp0);
+			break;
+		default:
+			return -1;
+	}
+
+	if (res)
+		*res = val;
+
+	return 0;
+}
+
+
+static int sdio_cmdSend(uint8_t cmd, uint32_t arg, uint32_t *res)
+{
+	int rslt;
+
+	mutexLock(sdio_common.cmd_lock);
+	rslt = sdio_cmdSendWorker(cmd, arg, res);
+	mutexUnlock(sdio_common.cmd_lock);
+
+	return rslt;
+}
+
+
+int sdio_transferDirect(sdio_dir_t dir, uint32_t address, uint8_t area,
+	uint8_t *data)
+{
+	int rslt = 0;
+	uint32_t response, arg = 0;
+
+	arg |= ((address & 0x1FFFF) << 9);
+	arg |= ((area & 0x7) << 28);
+	arg |= (dir << 31);
+
+	/* On direct write transfers data is part of argument */
+	if (dir == sdio_write) {
+		arg |= *data;
+	}
+
+	rslt = sdio_cmdSend(SDIO_CMD_RW_DIRECT, arg, &response);
+
+	/* Data is part of response on read transfers
+	 * and also on writes with RaW (read after write)
+	 * flag set - SDIO simplified spec v3.00, page 36 */
+	if (dir == sdio_read) {
+		*data = (uint8_t)(response & 0xFF);
+	}
+
+	return rslt;
+}
+
+
+int sdio_transferBulkWorker(sdio_dir_t dir, int blockMode, int opCode, uint32_t address,
```

Comment:
Do we need this `opCode` argument? whd always sets it to `1` when issuing command 53.

Reply:
Removed.